plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt

```diff
@@ -0,0 +1,177 @@
+    visit(JvmAnnotationNames.BYTECODE_VERSION_FIELD_NAME, header.bytecodeVersion)
+    visitArray(JvmAnnotationNames.METADATA_DATA_FIELD_NAME).apply {
+        header.data1.forEach { visit(null, it) }
+        visitEnd()
+    }
+    visitArray(JvmAnnotationNames.METADATA_STRINGS_FIELD_NAME).apply {
+        header.data2.forEach { visit(null, it) }
+        visitEnd()
+    }
+    visit(JvmAnnotationNames.METADATA_EXTRA_STRING_FIELD_NAME, header.extraString)
+    visit(JvmAnnotationNames.METADATA_PACKAGE_NAME_FIELD_NAME, header.packageName)
+    visit(JvmAnnotationNames.METADATA_EXTRA_INT_FIELD_NAME, header.extraInt)
+    visitEnd()
+}
+
+/**
+ * Class metadata adapter which removes private functions, properties, type aliases,
+ * and local delegated properties.
+ */
+private class AbiKmClassVisitor(delegate: KmClassVisitor) : KmClassVisitor(delegate) {
+    override fun visitFunction(flags: Flags, name: String): KmFunctionVisitor? {
+        if (!isPrivateDeclaration(flags))
+            return super.visitFunction(flags, name)
+        return null
+    }
+
+    override fun visitProperty(flags: Flags, name: String, getterFlags: Flags, setterFlags: Flags): KmPropertyVisitor? {
+        if (!isPrivateDeclaration(flags))
+            return super.visitProperty(flags, name, getterFlags, setterFlags)
+        return null
+    }
+
+    override fun visitTypeAlias(flags: Flags, name: String): KmTypeAliasVisitor? {
+        if (!isPrivateDeclaration(flags))
+            return super.visitTypeAlias(flags, name)
+        return null
+    }
+
+    override fun visitExtensions(type: KmExtensionType): KmClassExtensionVisitor? {
+        val delegate = super.visitExtensions(type)
+        if (type != JvmClassExtensionVisitor.TYPE) return delegate
+        return object : JvmClassExtensionVisitor(delegate as JvmClassExtensionVisitor?) {
+            override fun visitLocalDelegatedProperty(
+                flags: Flags, name: String, getterFlags: Flags, setterFlags: Flags
+            ): KmPropertyVisitor? = null
+        }
+    }
+}
+
+/**
+ * Class metadata adapter which removes private functions, properties, and type aliases.
+ */
+private class AbiKmPackageVisitor(delegate: KmPackageVisitor) : KmPackageVisitor(delegate) {
+    override fun visitFunction(flags: Flags, name: String): KmFunctionVisitor? {
+        if (!isPrivateDeclaration(flags))
+            return super.visitFunction(flags, name)
+        return null
+    }
+
+    override fun visitProperty(flags: Flags, name: String, getterFlags: Flags, setterFlags: Flags): KmPropertyVisitor? {
+        if (!isPrivateDeclaration(flags))
+            return super.visitProperty(flags, name, getterFlags, setterFlags)
+        return null
+    }
+
+    override fun visitTypeAlias(flags: Flags, name: String): KmTypeAliasVisitor? {
+        if (!isPrivateDeclaration(flags))
+            return super.visitTypeAlias(flags, name)
+        return null
+    }
+}
```

Comment:
Shouldn't metadata of local delegated properties declared in top level functions be removed here, similarly to how it's being removed for classes in `AbiKmClassVisitor`?

Reply:
Right. It's just that `JvmPackageExtensionVisitor` does not seem to have any methods for local delegated properties. Are we even recording top-level local delegated properties in the metadata?

Reply:
`JvmPackageExtensionVisitor` extends from `JvmDeclarationContainerExtensionVisitor` (like `JvmClassExtensionVisitor`), which declares `visitLocalDelegatedProperty`, so exactly the same approach with overriding `visitExtensions` should work

Reply:
Wow, I have no idea how I kept missing this. Thanks for pointing it out. It's fixed now.